### PR TITLE
Toggle transcripts and visible transcripts update

### PR DESCRIPTION
**Pull request by msmanoj**
## Description
This PR exposes the following mock APIs:
- Incoming action:
   - VISIBLE_TRANSCRIPTS - report from GB to inform the transcripts that are currently visible
- Outgoing action:
   - SET_VISIBLE_TRANSCRIPTS - used to instruct the GB to toggle the visibility of transcripts

I also removed `ZMENU_REPOSITION` actions as it is no longer used.

Related JIRA:
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1599

### CODE REVIEW

File: src/methods/handleIncoming.ts [hunk 1]
```diff
@@ -7,23 +7,33 @@ const handleIncoming = (subscriptions: Subscriptions, ...action: [type: Incoming
...
-    }
-    subscription(formattedIncoming);
+
+    [...subscriptionsToAction.values()].forEach(subscription => {
+
+      const formattedIncoming = formatIncoming(type, payload);
+      if (!formattedIncoming) {
+        console.error(payload);
+        return;
+      }
```
Comment (azangru) [inline src/methods/handleIncoming.ts]: out of curiosity, why do you repeatedly format the incoming message inside the forEach loop? What if you formatted it once outside the loop? And didn't call the subscribed functions if there is no formattedIncoming?

File: src/methods/send.ts [hunk 2]
```diff
@@ -73,6 +73,15 @@ const send = async (genomeBrowser: GenomeBrowserType, action: OutgoingAction) =>
     for(const track_id of action.payload.track_ids) {
       genomeBrowser.clear_switch(["track",track_id,"several"])
     }
+<<<<<<< HEAD
```
Comment (azangru) [inline src/methods/send.ts]: Tut-tut; git shouldn’t have allowed you to commit a conflict

File: src/types.ts [hunk 3]
```diff
@@ -14,6 +14,8 @@ export enum OutgoingActionType {
   TURN_OFF_NAMES = 'turn_off_names',
   TURN_ON_SEVERAL_TRANSCRIPTS = 'turn_on_several_transcripts',
   TURN_OFF_SEVERAL_TRANSCRIPTS = 'turn_off_several_transcripts',
+  TURN_ON_TRANSCRIPTS = 'turn_on_transcripts',
+  TURN_OFF_TRANSCRIPTS = 'turn_off_transcripts',
```
Comment (azangru) [inline src/types.ts]: Why do we have this pair of actions again? I thought the idea is that you are always telling the genome browser the list of transcript ids that you want to be visible, in which case the "on" and "off" actions are identical.

File: src/types.ts [hunk 4]
```diff
@@ -28,7 +30,8 @@ export enum IncomingActionType {
   SCROLL_POSITION = 'scroll_position',
   TRACK_SUMMARY = 'track_summary',
   ZMENU_CREATE = 'zmenu',
-  ZMENU_REPOSITION = 'update_zmenu_position'
+  ZMENU_REPOSITION = 'update_zmenu_position',
```
Comment (azangru) [inline src/types.ts]: Not related to the scope to this PR; but could you check with Dan if he thinks this action is going to be useful at all?

File: src/methods/send.ts [hunk 5]
```diff
@@ -73,13 +73,10 @@ const send = async (genomeBrowser: GenomeBrowserType, action: OutgoingAction) =>
     for (const track_id of action.payload.track_ids) {
       genomeBrowser.clear_switch(["track", track_id, "several"])
     }
-  } else if (action.type === OutgoingActionType.TURN_ON_TRANSCRIPTS) {
+  } else if (action.type === OutgoingActionType.SET_VISIBLE_TRANSCRIPTS) {
     const { track_id, transcript_ids } = action.payload;
-    genomeBrowser.set_switch(["track", track_id, "transcript", transcript_ids.join(" ")]);
+    genomeBrowser.set_switch(["track", track_id, "transcripts", transcript_ids.toString()]);
```
Comment (azangru) [inline src/methods/send.ts]: Not sure I am reading it correctly, but didn't Dan suggest passing an actual array? I.e. `JSON.stringify(transcript_ids)`?

File: src/methods/handleIncoming.ts (hunk 1 above)
Comment (msmanoj) [reply src/methods/handleIncoming.ts]: That's done

File: src/methods/handleIncoming.ts [hunk 6]
```diff
@@ -7,23 +7,33 @@ const handleIncoming = (subscriptions: Subscriptions, ...action: [type: Incoming
...
 
+  // Track summary payload also has the details about visible transcripts
+  // which we will be treating as a separate incoming action of its own.
+  if (type === IncomingActionType.TRACK_SUMMARY) {
+    handleIncoming(
+      subscriptions,
+      IncomingActionType.VISIBLE_TRANSCRIPTS,
+      payload
+    )
+  }
```
Comment (azangru) [inline src/methods/handleIncoming.ts:23]: Wouldn't you want to modify the payload when doing this? If I am reading this correctly, then what it means for ensembl-client is that the subscriber to the `VISIBLE_TRANSCRIPTS` message will have to deal with all track summaries, when it's really only interested in the ids of the visible transcripts.

_(We would also need some indication in the future that the focus track in TRACK_SUMMARY payload is actually a gene track, and thus actually contains transcript ids.)_

Comment (msmanoj) [reply src/methods/handleIncoming.ts:23]: Yes, but I thought it can be done when we have the actual payload ready.

Comment (azangru) [reply src/methods/handleIncoming.ts:23]: 👌

**Review by azangru — APPROVED** [review]

**Pull request merged**